src/csharp/BUILD-INTEGRATION.md

```diff
@@ -105,6 +105,17 @@ button](https://stackoverflow.com/a/9770061). Click on it, and choose "Add as
 link". If you do not select this option, Visual Studio will copy files to the
 project directory instead.
 
+#### My .proto files have same filename in different folders
+
+By default Grpc.Tools compiles all .proto files into `obj` directory, flattening relative paths. For proto files with duplicated names it will cause following error `NETSDK1022 Duplicate 'Compile' items were included. [...]`. If you want to keep relative paths in your `obj` folder, define `OutputDir="$(Protobuf_OutputPath)%(RelativeDir)"`.
+
+```xml
+  <ItemGroup>
+    <Protobuf Include="Protos/v2/http.proto" OutputDir="$(Protobuf_OutputPath)%(RelativeDir)"  />
```

Comment:
Based on my experiment, combining files from outside of the project's directory and specifying both `Link` and `OutputDir` doesn't seem to have the desired effect:
`<Protobuf Include="../../../protos/helloworld.proto" Link="helloworld.proto" OutputDir="$(Protobuf_OutputPath)%(RelativeDir)" />`  doesn't seem to have a good result because the RelativeDir value seems connected to the `Include` path, not the `Link` path. Perhaps in this case we need to use a different variable?

Reply:
Btw, there's also a %(Directory) variable that behaves slightly differently, but might be useful for out-of-project files
https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-well-known-item-metadata?view=vs-2019

Reply:
Using <Protobuf Include="../../../protos/bar/helloworld.proto" Link="someprefix/helloworld.proto" OutputDir="$(Protobuf_OutputPath)%(Link)" /> also kind of works except the `Link` metadata will be expanded to `someprefix/helloworld.proto`, so the generated code will go under `someprefix/helloworld.proto/Helloworld.cs` (dirty but works).

Reply:
Two questions came to my mind:
- Can you tell me what was the result of using `%(Directory)` instead of `%(RelativeDir)`? 
- What would be the satisfying result for Link (out of project) scenario?

Reply:
- It least on linux %(Directory) gave me the `OutputPath/the/full/path/without/the/root/Helloworld.cs` (I tried on linux), which worked but could cause problems with too long file names.

- It think any directory structure that works and still allows some reasonable way of seeing which .proto file corresponds to which .cs file would work. (actually I think our best way forward to solving this problem once and for all is what is proposed here: https://github.com/grpc/grpc/issues/17672#issuecomment-621152800)

Reply:
**TL;DR; we need stable hash mechanism, to avoid re-generating folders/files every build.**

I have tried using hash as output build as my early approach, my tries were based on mixing MSBuild well-known properties with MSBuild inline C# functions (not sure if this is the correct name for the latter). It wasn't successful because I was not able to pass value of Include to those functions. It doesn't really matter right now. 

Tip I would give to anyone that is going to implement this feature would be to have stable hash mechanism. I have had used .NET hash function and it generated, a new hash for the same object every application run. It is a new behaviour since .NET Core 3.x. 

If we don't cover this properly, we would generate tons of rubbish output every code generation.

Reply:
The approach from https://github.com/grpc/grpc/issues/17672#issuecomment-621152800  actually uses a prefix of a SHA1 hash (which is stable) and uses the full build path as input for the hash function (which is stable as well for any given file).

Reply:
@jtattermusch, do you need an extra hand to have it coded? If so, please leave some guideline were to start.

Reply:
I think @Falco20019  was going to implement something, but if you want to help out, it's always welcome.
Here's my experiment with the directory path hashing I'm currently testing: https://github.com/grpc/grpc/compare/master...jtattermusch:grpc_tools_fixes?expand=1

Reply:
Going to implement it today.